<div align="center">

# MEGR7172/8172

### Computational Methods (Fall 2025)
### Duke 227, Tu/Th 08:30 - 09:45 pm

</div>

# 7. Linear Algebra 1
Linear algebra is a branch of mathematics that studies vectors, vector spaces, linear transformations, and systems of linear equations. It is foundational for many areas in science and engineering, including data science, machine learning, physics, and computer graphics.

## 7.1 Scalars, Vectors, and Matrices
- *Scalar*: A single number.  
- *Vector*: An ordered list of numbers, often representing a point or direction in space.  
- *Matrix*: A rectangular array of numbers, representing a linear transformation or a system of equations.

In [5]:
import numpy as np

# Scalar: a single number
scalar = 5
print("Scalar:", scalar)

# Vector: a 1-dimensional array of numbers
vector = np.array([2, 4, 6])
print("Vector:", vector)

# Matrix: a 2-dimensional array of numbers
matrix = np.array([[1, 2], [3, 4]])
print("Matrix:\n", matrix)

Scalar: 5
Vector: [2 4 6]
Matrix:
 [[1 2]
 [3 4]]


## 7.2 Vector Operations

Vector operations are fundamental in linear algebra and include:

- **Addition:** Adding two vectors of the same size element-wise.
- **Subtraction:** Subtracting one vector from another element-wise.
- **Scalar Multiplication:** Multiplying each element of a vector by a scalar.
- **Dot Product:** The sum of the products of corresponding elements of two vectors. It results in a scalar.

Example using the `vector` variable:
- Addition: `vector + np.array([1, 1, 1])`
- Scalar multiplication: `2 * vector`
- Dot product: `np.dot(vector, np.array([1, 0, 1]))`

In [ ]:
vec_a = np.random.random(5)
vec_b = np.random.random(5)

print('vec_a + vec_b', vec_a + vec_b)
print('vec_a - vec_b', vec_a - vec_b)
print('2 * vec_a', 2 * vec_a)
print('Dot product', np.dot(vec_a, vec_b), vec_a @ vec_b)

vec_a + vec_b [1.13336921 0.88089878 0.81358427 1.06272783 1.03309388]
vec_a - vec_b [-0.39169593 -0.87733695  0.77123808 -0.50281377  0.10832391]
2 * vec_a [0.74167328 0.00356183 1.58482234 0.55991406 1.14141779]
Dot product 0.784147857235728


Questions:
1. What is the geometric interpretation of dot product between two vectors?
2. Do you have any other kinds of multiplications?
    - **Cross Product:** For 3D vectors, the cross product produces a vector perpendicular to both input vectors, e.g., `np.cross(a, b)`
    - **Element-wise Product:** Multiplies corresponding elements of two vectors or matrices, e.g., `a * b`.
    - **Outer Product:** Produces a matrix from two vectors, where each element is the product of elements from both vector, e.g., `np.outer(a, b)`.


## 7.3 Matrix and its Operations

A **matrix** is a rectangular array of numbers arranged in rows and columns. Matrices are used to represent linear transformations, systems of equations, and data in various fields.

**Common Matrix Operations:**

- **Addition/Subtraction:** Matrices of the same shape can be added or subtracted element-wise.
- **Scalar Multiplication:** Each element of a matrix is multiplied by a scalar.
- **Matrix Multiplication:** Combines two matrices to produce a new matrix. The number of columns in the first matrix must equal the number of rows in the second.
- **Transpose:** Flips a matrix over its diagonal, switching rows and columns.
- **Determinant:** A scalar value that can be computed from a square matrix, useful for solving systems and understanding invertibility.
- **Inverse:** The matrix that, when multiplied with the original, yields the identity matrix (only for square, non-singular matrices).

**Examples using the `matrix` variable:**
- Addition: `matrix + np.array([[5, 6], [7, 8]])`
- Scalar multiplication: `2 * matrix`
- Matrix multiplication: `matrix @ np.array([[5, 6], [7, 8]])`
- Transpose: `matrix.T`
- Determinant: `np.linalg.det(matrix)`
- Inverse: `np.linalg.inv(matrix)` (if the matrix is invertible)

In [20]:
# Example Matrix and operations
m1 = np.array([[1, 2], [3, 4]])
m2 = np.array([[5, 6], [7, 8]])
print("Matrix:\n", m1)

print("\nAddition:\n", m1 + m2)
print("\nScalar Multiplication:\n", 2 * m1)
print("\nMultiplication:\n", m1 @ m2)
print("\nTranspose:\n", m1.T)
print("\nDeterminant:", np.linalg.det(m1))
print("\nInverse:\n", np.linalg.inv(m1))


Matrix:
 [[1 2]
 [3 4]]

Addition:
 [[ 6  8]
 [10 12]]

Scalar Multiplication:
 [[2 4]
 [6 8]]

Multiplication:
 [[19 22]
 [43 50]]

Transpose:
 [[1 3]
 [2 4]]

Determinant: -2.0000000000000004

Inverse:
 [[-2.   1. ]
 [ 1.5 -0.5]]


Questions:
1. What is the geometric interpretation of determinant?
2. In addtion to the standard `m1 @ m2`, one also uses `m1 * m2`. Check the differences and list the application secenarios for both cases. 


## 7.4 Eigenvalues and Eigenvectors

**Eigenvalues** and **eigenvectors** are fundamental concepts in linear algebra, especially in the study of linear transformations and matrices.

- **Eigenvector:** A nonzero vector that only changes by a scalar factor when a linear transformation is applied.
- **Eigenvalue:** The scalar factor by which the eigenvector is scaled.

Mathematically, for a square matrix $A$, if there exists a vector $v$ and a scalar $\lambda$ such that:
$$
A v = \lambda v
$$
then $v$ is an eigenvector of $A$, and $\lambda$ is the corresponding eigenvalue.

**Applications:**
- Principal Component Analysis (PCA) in data science
- Stability analysis in engineering
- Quantum mechanics
- Google's PageRank algorithm

In [22]:
eigvals, eigvecs = np.linalg.eig(m1)
print("Eigenvalues:", eigvals)
print("Eigenvectors:\n", eigvecs)

Eigenvalues: [-0.37228132  5.37228132]
Eigenvectors:
 [[-0.82456484 -0.41597356]
 [ 0.56576746 -0.90937671]]


Questions:
1. Compute the eigenvalues for a matrix of $m1$:

    The eigenvalues of $m1 = \begin{bmatrix} 1 & 2 \\ 3 & 4 \end{bmatrix}$ are:
    $$
    \text{eigvals} = [-0.372, \; 5.372]
    $$

2. Explain the geometric meaning of eigenvalue and eigenvectors:

    - **Eigenvectors** indicate directions in which a linear transformation acts by stretching or compressing, but not rotating.
    - **Eigenvalues** represent the factor by which the eigenvector is scaled during the transformation.
    - Geometrically, applying the matrix to its eigenvector results in a new vector pointing in the same direction, scaled by the eigenvalue.